lib/common/xml.c

```diff
@@ -1203,45 +1204,46 @@ write_xml_stream(const xmlNode *xml, const char *filename, FILE *stream,
                  bool compress, unsigned int *nbytes)
 {
     // @COMPAT Drop nbytes as arg when we drop write_xml_fd()/write_xml_file()
-    int rc = pcmk_rc_ok;
-    unsigned int bytes_out = 0;
     char *buffer = NULL;
+    unsigned int bytes_out = 0;
+    int rc = pcmk_rc_ok;
 
     crm_log_xml_trace(xml, "writing");
 
     buffer = dump_xml_formatted(xml);
-    CRM_CHECK(buffer && strlen(buffer),
-              crm_log_xml_warn(xml, "formatting failed");
+    CRM_CHECK(!pcmk__str_empty(buffer),
+              crm_log_xml_info(xml, "dump-failed");
               rc = pcmk_rc_error;
               goto bail);
 
     if (compress) {
         unsigned int bytes_in = 0;
-        BZFILE *bz_file = NULL;
 
-        rc = BZ_OK;
-        bz_file = BZ2_bzWriteOpen(&rc, stream, 5, 0, 30);
-        rc = pcmk__bzlib2rc(rc);
+        // (5, 0, 0): (intermediate block size, silent, default workFactor)
+        BZFILE *bz_file = BZ2_bzWriteOpen(&rc, stream, 5, 0, 0);
 
-        if (rc != pcmk_rc_ok) {
+        if (rc != BZ_OK) {
+            rc = pcmk__bzlib2rc(rc);
```

Comment:
There are very specific codes that can be returned from each one of these libbz calls. Of those, there are very specific codes after which we're allowed to proceed instead of bailing out. I'm checking the ones that are acceptable in each context. For example, in one context `BZ_OK` and `BZ_STREAM_END` are the two "OK" values.

Reference:
* https://sourceware.org/bzip2/manual/manual.html

Reply:
See "3.4. High-level interface" and the various "Possible assignments to bzerror" and "Allowable next actions" blocks.

Reply:
Yep, the whole point was to avoid having to pore through bzip documentation whenever we need to modify code around these calls. The conversion provides a degree of insulation so we don't have to care.

Reply:
The code around these calls doesn't care except for the "OK to proceed" checks, which have no reason to change even if "what we proceed TO" changes.

It's not worth quibbling over though, so I'll revert to doing the conversion unconditionally

Reply:
We actually have to check the return codes correctly (meaning the way the bzip2 documentation tells us to) in at least one place:
https://github.com/ClusterLabs/pacemaker/blob/Pacemaker-2.1.7/lib/common/xml.c#L925-L937

Otherwise we can't tell whether we exited due to stream end or due to some error.

I'd prefer to be consistent, but could leave one section that checks the bzip2 error codes and let the rest check the mapped return code.

Reply:
Yeah I'd rather convert beforehand wherever possible. It makes sense to check the direct code in the one place.